### PR TITLE
HWKAPM-781 - Added support for OpenTracing

### PR DESCRIPTION
**Pull request by jpkrohling**

### CODE REVIEW

**Comment by jpkrohling** [comment]
@objectiser , @pavolloffay , @rafabene could you please review? Note that the changes from this PR should be merged around the same time as similar changes for the other modules.

**Comment by jpkrohling** [comment]
Sample screens after the changes to all modules:

https://snag.gy/1PHovA.jpg
https://snag.gy/GaKJFA.jpg
https://snag.gy/soW5ND.jpg

In APM terms, this is the data recorded per request:
https://paste.fedoraproject.org/519408/34607821/ ( http://jsonviewer.stack.hu/ might be a good tool to use for visualizing this)

File: src/main/java/com/redhat/developers/msa/api_gateway/GenericFeignClient.java [hunk 1]
```diff
@@ -78,18 +69,27 @@ public GenericFeignClient(Class<T> classType, String serviceName, T fallback, Br
      * 
      * @return The feign pointing to the service URL and with Hystrix fallback.
      */
-    protected T createFeign(ServerSpan serverSpan) {
+    protected T createFeign(Span serverSpan) {
+        System.out.println(String.format("Creating feign for service {%s} under {%s}", serviceName, serverSpan.toString()));
```
Comment (jpkrohling) [inline src/main/java/com/redhat/developers/msa/api_gateway/GenericFeignClient.java]: Remove sysouts

File: src/main/java/com/redhat/developers/msa/api_gateway/GenericFeignClient.java [hunk 2]
```diff
@@ -78,18 +69,27 @@ public GenericFeignClient(Class<T> classType, String serviceName, T fallback, Br
...
         final CloseableHttpClient httpclient =
             HttpClients.custom()
-                .addInterceptorFirst(new BraveHttpRequestInterceptor(brave.clientRequestInterceptor(), new DefaultSpanNameProvider()))
-                .addInterceptorFirst(new BraveHttpResponseInterceptor(brave.clientResponseInterceptor()))
+                .addInterceptorFirst(new ApiGatewayHttpRequestInterceptor(serverSpan))
+                .addInterceptorFirst(new ApiGatewayHttpResponseInterceptor(serverSpan))
                 .build();
-        String url = String.format("http://%s:8080/", serviceName);
+
+        String url = System.getenv(String.format("%s_SERVER_URL", serviceName.toUpperCase()));
```
Comment (jpkrohling) [inline src/main/java/com/redhat/developers/msa/api_gateway/GenericFeignClient.java:79]: Document this.

File: src/main/java/com/redhat/developers/msa/api_gateway/tracing/OpenTracingConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+ *
+ * @author Juraci Paixão Kröhling
+ */
+@SuppressWarnings("Duplicates")
+@Configuration
+public class OpenTracingConfiguration {
+
+    @Bean
+    public FilterRegistrationBean getOpenTracingFilter() {
+        return new FilterRegistrationBean(new Filter() {
```
Comment (jpkrohling) [inline src/main/java/com/redhat/developers/msa/api_gateway/tracing/OpenTracingConfiguration.java:57]: Similar question as the one for `ola`: does Spring Boot get a bean per request, or per deployment?

Comment (pavolloffay) [reply src/main/java/com/redhat/developers/msa/api_gateway/tracing/OpenTracingConfiguration.java:57]: IIRC by default it's singleton. 

http://docs.spring.io/spring/docs/3.0.0.M3/reference/html/ch04s04.html

File: src/main/java/com/redhat/developers/msa/api_gateway/tracing/OpenTracingConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            @Override public void init(FilterConfig filterConfig) throws ServletException {
+            }
+
+            @Override public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain) throws IOException, ServletException {
```
Comment (pavolloffay) [inline src/main/java/com/redhat/developers/msa/api_gateway/tracing/OpenTracingConfiguration.java:61]: What is the execution of this? 

1. called  before processing in controller
2. stop at `filterChain.doFilter`
3. continue after processing controller?

**Review by pavolloffay — COMMENTED** [review]
LGTM :+1: 

I did not comment everything but apply the same things as in other modules:

* noop tracer, you can use also MockTracer and override ~onSpanFinished and send it to stdout.
* standard tags
* tracer null check
* `tracing.requestSpan` constant

File: src/main/java/com/redhat/developers/msa/api_gateway/tracing/OpenTracingConfiguration.java (hunk 4 above)
Comment (jpkrohling) [reply src/main/java/com/redhat/developers/msa/api_gateway/tracing/OpenTracingConfiguration.java:61]: It's standard servlet semantics: the call to `filterChain#doFilter()` will call other filters on the chain or will execute the target servlet if no other filters are available. So, we have the opportunity to have code executed before and after the target servlet. Is that your question?

**Comment by jpkrohling** [comment]
This PR has been updated. I believe this is now ready to be merged, but a review from @rafabene would be useful.

**Comment by jpkrohling** [comment]
By the way: the Brave OpenTracing library currently has a bug that prevents it from correctly sending data to ZipKin. @pavolloffay might be able to provide more details about it, and the library's version will be changed as soon as a fix is released.

**Comment by pavolloffay** [comment]
Do you mean that "bug" related to `MockSpanBuilder` being returned on `tracer#extract(type, carrier)` with carrier without span context? Their are currently expect this behaviour see https://github.com/openzipkin/brave-opentracing/blob/master/src/test/java/io/opentracing/impl/BraveTracerTest.java#L355. I have raised a question on gitter/zipkin I will let you know.

**Comment by jpkrohling** [comment]
The Noop, not the Mock, but yes :) Note that either APM is wrong, or Brave, as the application uses only OpenTracing APIs.

**Comment by pavolloffay** [comment]
I would say that Brave is wrong. You don't know whether there is span context in carrier or not. Tracer should definitely start the span.

**Comment by jpkrohling** [comment]
PR updated with the final changes.

File: src/main/java/com/redhat/developers/msa/api_gateway/tracing/OpenTracingConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                            .asChildOf(spanContext)
+                            .start();
+                    Tags.HTTP_URL.set(requestSpan, httpServletRequest.getRequestURI());
+                    servletRequest.setAttribute("tracing.requestSpan", requestSpan);
+                }
+
+                filterChain.doFilter(servletRequest, servletResponse);
+
+                if (null != requestSpan) {
+                    if (servletResponse instanceof HttpServletResponse) {
```
Comment (pavolloffay) [inline src/main/java/com/redhat/developers/msa/api_gateway/tracing/OpenTracingConfiguration.java:77]: nit: I think this is not necessary.

**Review by pavolloffay — APPROVED** [review]
LGTM :+1:

**Comment by jpkrohling** [comment]
This PR has been updated to remove APM. This is now only about adding support to OpenTracing, with ZipKin as the sole available implementation.

**Pull request closed without merging**